### PR TITLE
Chapter 6: Modeling users

### PR DESCRIPTION
**Pull request by baaohuii**

### CODE REVIEW

**Comment by baaohuii** [comment]
ready

File: app/models/user.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+class User < ApplicationRecord
+  before_save { self.email = email.downcase }
+  validates :name, presence: true, length: { maximum: 50 }
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+(\.[a-z\d\-]+)*\.[a-z]+\z/i
+  validates :email, presence: true, length: {maximum: 50 }, 
+                    format: { with: VALID_EMAIL_REGEX }, 
+                    uniqueness: { case_sensitive: true }
+  has_secure_password
```
Comment (saobv-2223) [inline app/models/user.rb]: move this line below validate

File: app/models/user.rb [hunk 2]
```diff
@@ -0,0 +1,10 @@
+class User < ApplicationRecord
+  before_save { self.email = email.downcase }
+  validates :name, presence: true, length: { maximum: 50 }
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+(\.[a-z\d\-]+)*\.[a-z]+\z/i
```
Comment (saobv-2223) [inline app/models/user.rb]: move constant to top

File: app/models/user.rb [hunk 3]
```diff
@@ -0,0 +1,10 @@
+class User < ApplicationRecord
+  before_save { self.email = email.downcase }
```
Comment (saobv-2223) [inline app/models/user.rb]: move before save below has_secure_password

**Comment by baaohuii** [comment]
ready

File: app/models/user.rb [hunk 4]
```diff
@@ -0,0 +1,10 @@
+class User < ApplicationRecord
+  VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+(\.[a-z\d\-]+)*\.[a-z]+\z/i
+  validates :name, presence: true, length: { maximum: 50 }
+  validates :email, presence: true, length: {maximum: 50 }, 
+                    format: { with: VALID_EMAIL_REGEX }, 
+                    uniqueness: { case_sensitive: true }
+  validates :password, presence: true, length: { minimum: 6 }
+  has_secure_password
+  before_save { self.email = email.downcase }
```
Comment (saobv-2223) [inline app/models/user.rb]: above and below line 8, you should have an empty line

Comment (saobv-2223) [inline app/models/user.rb]:
```suggestion
  validates :password, presence: true, length: { minimum: 6 }
  
  has_secure_password
  
  before_save downcase_email
  
  private
  
  downcase_email
    email.downcase!  
  end
```

**Comment by baaohuii** [comment]
ready

File: app/models/user.rb [hunk 5]
```diff
@@ -0,0 +1,17 @@
...
+  validates :password, presence: true, length: { minimum: 6 }
+
+  has_secure_password
+
+  before_save downcase_email
+  
+  private
+  
+  downcase_email
+    email.downcase!
```
Comment (saobv-2223) [inline app/models/user.rb]:
```suggestion
  downcase_email
    email.downcase!  
  end
```

**Comment by baaohuii** [comment]
ready

**Comment by manhnt-0960** [comment]
merged

**Pull request merged**